### ISSUE TITLE
SparkSQL to ES invalid "IN" query   

### ISSUE THREAD

**Issue by alexbudniy** [elastic/elasticsearch-hadoop#556]
In short:

Using maven dependencies:
```xml
<dependency>
  <groupId>org.elasticsearch</groupId>
 <artifactId>elasticsearch-spark_2.10</artifactId>
 <version>2.1.1</version>
</dependency>

<dependency>
  <groupId>org.apache.spark</groupId>
  <artifactId>spark-core_2.10</artifactId>
  <version>1.5.0</version>
</dependency>
```

The following query
```
sql.sql("select mac from sessionTable where hotspot_id in (1, 2, 3)");
```
is translated into a wrong ES query causing an error:
```
[Failed to parse source [{"query":{"filtered":{ "query":{"match_all":{}},"filter": { "and" : [ {"query":{"match":{"hotspot_id":"1 2 3"}}} ] } }}}]]]; nested: NumberFormatException[For input string: "1 2 3"];
```

The field type:
"hotspot_id": {"type": "long"}

Full code:
```
SparkConf conf = new SparkConf().setAppName("SparkEsTest").setMaster("local[2]");
conf.set("es.nodes", "localhost");
conf.set("es.nodes.client.only", "false");
conf.set("es.scroll.size", "10000");
JavaSparkContext sc = new JavaSparkContext(conf);

SQLContext sql = new SQLContext(sc);
DataFrame dataFrame = sql.read().format("org.elasticsearch.spark.sql").load("session/session");
dataFrame.registerTempTable("sessionTable");
DataFrame data = sql.sql("select mac from sessionTable where hotspot_id in (1, 2, 3)");
data.collect();
```

**Comment by alexbudniy** [comment]
Oh, after looking at sources I found a fix - "IN" query is properly generated if the following property is set:
conf.set("es.internal.spark.sql.pushdown.strict", "true");

Closing the issue.

**Comment by costin** [comment]
Reopening this issue. `es.internal.` is reserved for internal configuration and it's not supported. If I understand correctly when you enable strict pushdown the filter is properly created. However the default is not strict (meaning it's working on analyzed vs non-analyzed filters).

Long story short - you can change the strict aspect through a proper setting that will be portable across settings. Second, when not using a strict pushdown, the query appears to be invalid which is a bug.

**Comment by alexbudniy** [comment]
Ok, in this case the problematic code is most likely the following:

sql-13, class org.elasticsearch.spark.sql.DefaultSource.scala, line 155-158:

```
case In(attribute, values)                => {
    if (strictPushDown) s"""{"terms":{"$attribute":${extractAsJsonArray(values)}}}"""
    else s"""{"query":{"match":{"$attribute":${extract(values)}}}}"""
}
```
if strictPushDown is true it uses "terms" correctly.

**Comment by costin** [comment]
This has been fixed in master and 2.1. It was a bit more difficult as the type influenced the type of the query generated.

Thanks for the report!
